bpf/lib/nodeport.h

```diff
@@ -1946,7 +1946,8 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, int *ifindex
 	l4_off = l3_off + ipv4_hdrlen(ip4);
 	csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
 
-#if defined(ENABLE_EGRESS_GATEWAY) && !defined(TUNNEL_MODE)
+#if defined(ENABLE_EGRESS_GATEWAY) && !defined(TUNNEL_MODE) && \
+	__ctx_is != __ctx_xdp
```

Comment:
Could we do this from agent side when we prepare bpf_xdp.c vs bpf_host, so that `__ctx_is != __ctx_xdp` is not needed given for XDP `ENABLE_EGRESS_GATEWAY`, `TUNNEL_MODE` is not defined?

Reply:
The `ENABLE_EGRESS_GATEWAY` is defined for bpf_xdp. With the proposed changes, bpf_xdp will do the rev SNAT for the egress GW reply traffic. The only missing piece is doing redirecting to the tunnel netdev / doing the encap.

If we undef `ENABLE_EGRESS_GATEWAY` for bpf_xdp, then the reply packet won't be handled at all due to `XFER_PKT_NO_SVC` set by bpf_xdp.

Reply:
Good point, this detail should probably be commented somewhere. The code looks contained enough to refactor into its own function (e.g. for [egress_policies.h](https://github.com/cilium/cilium/blob/master/bpf/lib/egress_policies.h)) so we wouldn't need the extra ifdefs in here?